lib/icingadb/icingadb-objects.cpp

```diff
@@ -1383,8 +1383,11 @@ bool IcingaDB::PrepareObject(const ConfigObject::Ptr& object, Dictionary::Ptr& a
 			attributes->Set("timeperiod_id", GetObjectIdentifier(timeperiod));
 
 		if (notification->GetTimes()) {
-			attributes->Set("times_begin", notification->GetTimes()->Get("begin"));
-			attributes->Set("times_end",notification->GetTimes()->Get("end"));
+			auto begin (notification->GetTimes()->Get("begin"));
+			auto end (notification->GetTimes()->Get("end"));
+
+			attributes->Set("times_begin", begin.GetType() == ValueEmpty ? begin : Value(std::round((double)begin)));
+			attributes->Set("times_end", end.GetType() == ValueEmpty ? end : Value(std::round((double)end)));
```

Comment:
That check looks oddly specific. Would anything be wrong with something simpler like this?
```c++
begin.IsEmpty() ? Empty : Value(std::round((double)begin))
```

However, one more thing: when `times.begin`/`times.end` is set to a negative value, it's basically treated like `null`:
https://github.com/Icinga/icinga2/blob/5350aa3c72c58a66d56561cdd62bee5fe9d74485/lib/icinga/notification.cpp#L286
https://github.com/Icinga/icinga2/blob/5350aa3c72c58a66d56561cdd62bee5fe9d74485/lib/icinga/notification.cpp#L305

So I think it would make sense to only conditionally set these values with a similar condition. That would also properly handles negative values which are problematic because the column type in the schema is actually unsigned (couldn't get it to crash with MariaDB though, it just sets the value to 0, but I guess PostgreSQL would be more strict).

Reply:
> couldn't get it to crash with MariaDB though, it just sets the value to 0

I'm not quite sure.

```
MariaDB [idb]> create table lolcat ( u int unsigned );
Query OK, 0 rows affected (0,046 sec)

MariaDB [idb]> insert into lolcat values(-1);
ERROR 1264 (22003): Out of range value for column 'u' at row 1
MariaDB [idb]>
```

Reply:
> but I guess PostgreSQL would be more strict

👍

```
2023-06-27T11:26:47.654+0200	WARN	config-sync	Aborted initial state sync after 198.785737ms
2023-06-27T11:26:47.780+0200	WARN	config-sync	Aborted config sync after 324.62004ms
2023-06-27T11:26:47.780+0200	FATAL	icingadb	pq: value for domain uint violates check constraint "between_0_and_4294967295"
can't perform "INSERT INTO \"notification\" (\"name\", \"name_checksum\", \"id\", \"notificationcommand_id\", \"times_end\", \"notification_interval\", \"environment_id\", \"name_ci\", \"properties_checksum\", \"host_id\", \"times_begin\", \"service_id\", \"timeperiod_id\", \"states\", \"types\", \"zone_id\") VALUES (:name, :name_checksum, :id, :notificationcommand_id, :times_end, :notification_interval, :environment_id, :name_ci, :properties_checksum, :host_id, :times_begin, :service_id, :timeperiod_id, :states, :types, :zone_id)"
github.com/icinga/icingadb/internal.CantPerformQuery
	/Users/aklimov/NET/WS/icingadb/internal/internal.go:30
```

Reply:
Mystery solved: https://github.com/Icinga/icingadb/issues/611

```
MariaDB [icingadb]> SET sql_mode = 'ANSI_QUOTES';
Query OK, 0 rows affected (0.000 sec)

MariaDB [icingadb]> create table lolcat ( u int unsigned );
Query OK, 0 rows affected (0.045 sec)

MariaDB [icingadb]> insert into lolcat values(-1);
Query OK, 1 row affected, 1 warning (0.004 sec)

MariaDB [icingadb]> select * from lolcat;
+------+
| u    |
+------+
|    0 |
+------+
1 row in set (0.001 sec)
```